### PR TITLE
Add feature to convert from ReferenceEquals to 'is null'

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide   Tests forthcoming

**Comment by CyrusNajmabadi** [comment]
TODO: add language version check.
Add user facing option.

**Comment by alrz** [comment]
Perhaps `(object)obj ==/!= null` or a null check where `==` or `!=` are not user defined, are good candidates.

**Comment by CyrusNajmabadi** [comment]
@alrz I think that's fine to add as another option

File: src/EditorFeatures/CSharpTest/UseIsNullCheck/UseIsNullCheckTests.cs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+@"using System;
+
+class C
+{
+    void M(string s)
+    {
+        if ([||]ReferenceEquals(s, Nothing))
+            return;
+    }
+end class",
```
Comment (Therzok) [inline src/EditorFeatures/CSharpTest/UseIsNullCheck/UseIsNullCheckTests.cs]: VB syntax?

File: src/EditorFeatures/CSharpTest/UseExpressionBody/Analyzer/UseExpressionBodyForAccessorsAnalyzerTests.cs [hunk 2]
```diff
@@ -52,14 +52,14 @@ int Foo
             [|return|] Bar();
         }
     }
-}",
+end class",
```
Comment (Therzok) [inline src/EditorFeatures/CSharpTest/UseExpressionBody/Analyzer/UseExpressionBodyForAccessorsAnalyzerTests.cs]: VB syntax in this file too?

File: src/Workspaces/VisualBasicTest/CodeGeneration/SyntaxGeneratorTests.vb [hunk 3]
```diff
@@ -332,11 +332,11 @@ End Class
         Public Sub TestArrayCreationExpressions()
             VerifySyntax(Of ArrayCreationExpressionSyntax)(
                 _g.ArrayCreationExpression(_g.IdentifierName("x"), _g.LiteralExpression(10)),
-                "New x(10) {}")
+                "New x(10) {end class")
```
Comment (Therzok) [inline src/Workspaces/VisualBasicTest/CodeGeneration/SyntaxGeneratorTests.vb]: Probably a mishap here too and `zend class` below

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs [hunk 4]
```diff
@@ -121,63 +121,63 @@ public void TestAttributeData()
         {
             VerifySyntax<AttributeListSyntax>(_g.Attribute(GetAttributeData(
 @"using System; 
-public class MyAttribute : Attribute { }",
+public class MyAttribute : Attribute { end class",
```
Comment (Therzok) [inline src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs]: `end class` here too

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Preemptively merging.  Let me know if there's anything you want me to change.

File: src/EditorFeatures/CSharpTest/UseIsNullCheck/UseIsNullCheckTests.cs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            return;
+    }
+}",
+@"using System;
+
+class C
+{
+    void M(string s)
+    {
+        if (!(s is null))
```
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/UseIsNullCheck/UseIsNullCheckTests.cs]: This makes me wish C# had iSnot.

File: src/Features/Core/Portable/UseIsNullCheck/AbstractUseIsNullDiagnosticAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(compilationContext => {
```
Comment (Pilchie) [inline src/Features/Core/Portable/UseIsNullCheck/AbstractUseIsNullDiagnosticAnalyzer.cs]: I don't like => followed by a complex lambda.

File: src/Features/Core/Portable/UseIsNullCheck/AbstractUseIsNullDiagnosticAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            {
+                return;
+            }
+
+            var symbol = semanticModel.GetSymbolInfo(invocation, cancellationToken).Symbol;
+            if (!referenceEqualsMethod.Equals(symbol))
+            {
+                return;
+            }
+
```
Comment (Pilchie) [inline src/Features/Core/Portable/UseIsNullCheck/AbstractUseIsNullDiagnosticAnalyzer.cs]: Extra newline.

File: src/Features/Core/Portable/UseIsNullCheck/AbstractUseIsNullDiagnosticAnalyzer.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+            var additionalLocations = ImmutableArray.Create(invocation.GetLocation());
+            var properties = ImmutableDictionary<string, string>.Empty;
+
+            var negated = syntaxFacts.IsLogicalNotExpression(invocation.Parent);
```
Comment (Pilchie) [inline src/Features/Core/Portable/UseIsNullCheck/AbstractUseIsNullDiagnosticAnalyzer.cs]: Consider parens between?

Comment (Pilchie) [inline src/Features/Core/Portable/UseIsNullCheck/AbstractUseIsNullDiagnosticAnalyzer.cs]: Consider parens between?

File: src/Features/VisualBasic/Portable/UseIsNullCheck/VisualBasicUseIsNullCheckCodeFixProvider.vb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+            Return VBFeaturesResources.use_IsNot_Nothing_check
+        End Function
+
+        Protected Overrides Function CreateIsNullCheck(argument As SyntaxNode) As SyntaxNode
+            Return SyntaxFactory.IsExpression(
+                DirectCast(argument, ExpressionSyntax).Parenthesize(),
+                SyntaxFactory.NothingLiteralExpression(SyntaxFactory.Token(SyntaxKind.NothingKeyword))).Parenthesize()
+        End Function
+
+        Protected Overrides Function CreateIsNotNullCheck(notExpression As SyntaxNode, argument As SyntaxNode) As SyntaxNode
```
Comment (Pilchie) [inline src/Features/VisualBasic/Portable/UseIsNullCheck/VisualBasicUseIsNullCheckCodeFixProvider.vb]: This is weird.  What happens to trivia, etc?

Comment (Pilchie) [inline src/Features/VisualBasic/Portable/UseIsNullCheck/VisualBasicUseIsNullCheckCodeFixProvider.vb]: Also, that `notExpression` is not used.

**Pull request merged**